python/ray/autoscaler/_private/aws/config.py

```diff
@@ -762,21 +767,158 @@ def _get_key(key_name, config):
         raise exc
 
 
-def _configure_from_network_interfaces(config):
+def _configure_from_launch_template(config: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Merges any launch template data referenced by the node config of all
+    available node type's into their parent node config. Any parameters
+    specified in node config override the same parameters in the launch
+    template, in compliance with the behavior of the ec2.create_instances
+    API. The config to bootstrap is modified in place.
+
+    Args:
+        config (Dict[str, Any]): config to bootstrap
+    Returns:
+        config (Dict[str, Any]): The input config with all launch template
+        data merged into the node config of all available node types. If no
+        launch template data is found, then the config is returned
+        unchanged.
+    Raises:
+        ValueError: If no launch template is found for any launch
+        template [name|id] and version, or more than one launch template is
+        found.
+    """
+
+    # iterate over sorted node types to support deterministic unit test stubs
+    for _, node_type in sorted(config["available_node_types"].items()):
+        config = _configure_node_type_from_launch_template(config, node_type)
+    return config
```

Comment:
Can we have this not modify the config in place?

Reply:
Done. 

Note that this also required patching a couple areas that depend on `bootstrap_aws` to invalidate their assumption of in-place config modification. I've also introduced a `copy.deepcopy(config)` line at the top of `bootstrap_aws` to both prevent in-place modification of the input config to bootstrap by any existing code path, and to make this behavior more immediately clear to readers.

Cleaning up in-place modification of each individual section of the copied config (e.g. provider config, node type config, node config, etc.) seems like a good thing to target across subsequent PRs.